rustler_codegen/src/tagged_enum.rs

```diff
@@ -0,0 +1,261 @@
+use std::collections::HashMap;
+
+use proc_macro2::{Span, TokenStream};
+use quote::{quote, quote_spanned};
+
+use heck::ToSnakeCase;
+use syn::{self, spanned::Spanned, Fields, Ident, Variant};
+
+use super::context::Context;
+
+pub fn transcoder_decorator(ast: &syn::DeriveInput) -> TokenStream {
+    let ctx = Context::from_ast(ast);
+
+    let variants = ctx
+        .variants
+        .as_ref()
+        .expect("NifEnum can only be used with enums");
+
+    // Remove duplicated atoms.
+    let atom_set = variants
+        .iter()
+        .flat_map(|variant| {
+            let mut ret: Vec<(String, Ident)> = if let Fields::Named(fields) = &variant.fields {
+                fields
+                    .named
+                    .iter()
+                    .map(|field| {
+                        let atom_str = field
+                            .ident
+                            .as_ref()
+                            .expect("Atom is expected")
+                            .to_string()
+                            .to_snake_case();
+                        let atom_fn = Ident::new(&format!("atom_{}", atom_str), Span::call_site());
+                        (atom_str, atom_fn)
+                    })
+                    .collect()
+            } else {
+                vec![]
+            };
+
+            let atom_str = variant.ident.to_string().to_snake_case();
+            let atom_fn = Ident::new(&format!("atom_{}", atom_str), Span::call_site());
+            ret.push((atom_str, atom_fn));
+
+            ret
+        })
+        .collect::<HashMap<_, _>>();
```

Comment:
I agree deduplicating is not needed. At first, I intended to prevent making the same atoms, but I think it makes codes more complicated and not actually necessary. 😃

Reply:
I think we don't need to test it if it already works.